File: javascript/easy/two_sum.js
/**
 * twoSum
 * @param { Array < number >} arr
 * @param { number; } target
 * 
 * @returns { Array < number >}
 */
function twoSum(arr, target) {
  const result = [];

  for (let i = 0; i < arr.length; i++) {
    const vInner = arr[i];
    for (let j = i + 1; j < arr.length; j++) {
      const vOuter = arr[j];
      if (vInner + vOuter === target) {
        result.push(i, j + i);
        return result;
      }
    }
  }

  return result;
}

/**
 * twoSum
 * @param { Array < number >} arr
 * @param { number; } target
 * 
 * @returns { Array < number >}
 */
function twoSum2(arr, target) {
  const map = {};

  for (let i = 0; i < arr.length; i++) {
    const v = arr[i];
    const diff = target - v;

    if (diff in map) {
      return [map[diff], i];
    }

    map[v] = i;
  }

  return [];
}

function main() {
  console.log(twoSum([2, 7, 11, 15], 9), [0, 1]);

  const start1 = new Date().getTime();
  const bigArr = [];

  for (let i = 0; i < 100_000_000; i++) {
    // get random number 1 - 1_000_000
    const rand = Math.floor(Math.random() * 1_000_000) + 1;
    bigArr.push(rand);
  }

  console.log(bigArr.length);
  const end1 = new Date().getTime();
  console.log(`Time: ${end1 - start1}ms`);


  const target = 1_500_000;

  // get Start timer
  const start = new Date().getTime();
  const [a, b] = twoSum2(bigArr, target);

  console.log(a, b, 'values', bigArr[a], bigArr[b], 'sum', bigArr[a] + bigArr[b], target);
  // get End timer
  const end = new Date().getTime();
  console.log(`Time: ${end - start}ms`);
}

main();

File: go/easy/two_sum.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func twoSum(nums []int, target int) []int {
	m := make(map[int]int, len(nums))
	for i, n := range nums {
		if j, ok := m[target-n]; ok {
			return []int{j, i}
		}
		m[n] = i
	}
	return nil
}

// using for range
func twoSum2(nums []int, target int) []int {
	result := []int{}

	for i, vOuter := range nums {
		for j, vInner := range nums[i+1:] {
			if vOuter+vInner == target {
				result = append(result, i, j+i+1)
				break
			}
		}

		if len(result) > 0 {
			break
		}
	}

	return result
}

func main() {
	fmt.Printf("%v %v\n", twoSum([]int{2, 7, 11, 15}, 9), []int{0, 1})
	fmt.Printf("%v %v\n", twoSum([]int{2, 7, 11, 15}, 18), []int{1, 2})

	bigArr := make([]int, 10_000_000)

	for i := 0; i < 10_000_000; i++ {
		bigArr[i] = int(rand.Int63n(int64(1_000_000)))
	}

	start := time.Now()
	fmt.Printf("big map %v %v\n", twoSum(bigArr, 1_500_000), []int{1, 2})
	fmt.Printf("big map %v\n", time.Since(start))

	start = time.Now()
	fmt.Printf("big for-range %v %v\n", twoSum2(bigArr, 1_500_000), []int{1, 2})
	fmt.Printf("big for-range %v\n", time.Since(start))
}
